### PR TITLE
fix: Clear bytecode for smart-contracts self destructed in a separate transaction

### PR DESCRIPTION
**Pull request by vbaranov**
## Motivation

Resolves https://github.com/blockscout/blockscout/issues/8923

### AI Agent Summary

This pull request introduces logic to empty the bytecode (`contract_code`) of smart contracts that have been selfdestructed, except for contracts that were created and selfdestructed within the same transaction. The changes include both runtime logic in the internal transaction import process and a new migration for existing data. Comprehensive tests are added to ensure correct behavior in various scenarios.

**Internal Transaction Import Logic:**

* Added a step to the internal transaction import pipeline to empty the `contract_code` for addresses that were selfdestructed, unless they were also created in the same transaction. This ensures that contract bytecode is removed only for contracts that truly no longer exist. [[1]](diffhunk://#diff-ac792f8a8aa0e6910e76fccd5228611c5758ef6e43bcf1771e01543d4eb70211R205-R215) [[2]](diffhunk://#diff-ac792f8a8aa0e6910e76fccd5228611c5758ef6e43bcf1771e01543d4eb70211R916-R962)
* Refactored to include the `Address` module in the relevant import for bytecode updates.

**Database Migration:**

* Introduced a new migration module, `Explorer.Migrator.EmptyBytecodeForSelfdestructedSmartContracts`, which processes historical data to empty `contract_code` for all previously selfdestructed contracts, again skipping those created and destroyed in the same transaction. This migration works in batches for efficiency.

**Testing Enhancements:**

* Added comprehensive tests covering:
  - Emptying bytecode for contracts selfdestructed in a different transaction from creation.
  - Not emptying bytecode when creation and selfdestruct happen in the same transaction (for both `create` and `create2`).
  - Handling of multiple contracts and non-existent addresses gracefully.

## Checklist for your Pull Request (PR)

- [ ] I verified this PR does not break any public APIs, contracts, or interfaces that external consumers depend on.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I updated documentation if needed:
  - [ ] General docs: submitted PR to [docs repository](https://github.com/blockscout/docs).
  - [ ] ENV vars: updated [env vars list](https://github.com/blockscout/docs/tree/main/setup/env-variables) and set version parameter to `master`.
  - [ ] Deprecated vars: added to [deprecated env vars list](https://github.com/blockscout/docs/tree/main/setup/env-variables/deprecated-env-variables).
- [ ] If I modified API endpoints, I updated the Swagger/OpenAPI schemas accordingly and checked that schemas are asserted in tests.
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Summary by CodeRabbit

* **New Features**
  * Smart contract bytecode is cleared for contracts that self-destruct in separate transactions; a migration was added to process historical cases.

* **Bug Fixes**
  * Bytecode is preserved when a contract is created and self-destructs within the same transaction; handles multiple contracts and non-existent addresses gracefully.

* **Tests**
  * Extensive tests added covering creation/self-destruct edge cases, multi-contract scenarios, and the migration behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by vbaranov** [comment]
@coderabbitai review

File: apps/explorer/lib/explorer/chain/import/runner/internal_transactions.ex [hunk 1]
```diff
@@ -901,6 +913,54 @@ defmodule Explorer.Chain.Import.Runner.InternalTransactions do
...
+    else
+      # Update the addresses to have empty contract_code
+      empty_contract_code = %Explorer.Chain.Data{bytes: <<>>}
+
+      update_query =
+        from(
+          address in Address,
+          where: address.hash in ^addresses_to_empty,
+          update: [set: [contract_code: ^empty_contract_code, updated_at: ^timestamps.updated_at]]
+        )
```
Comment (rimrakhimov) [inline apps/explorer/lib/explorer/chain/import/runner/internal_transactions.ex:952]: I'm not sure we would actually like to consider such a rare scenario, but what happens if self-desctruct transaction is reorged, so that the address will stay without bytecode even though the transaction was reverted in reality? Is it the case that our on-demand contract code mechanism will take place, and the corresponding bytecode will be restored anyway?

File: apps/explorer/lib/explorer/chain/import/runner/internal_transactions.ex [hunk 2]
```diff
@@ -901,6 +913,54 @@ defmodule Explorer.Chain.Import.Runner.InternalTransactions do
...
+      |> MapSet.new()
+
+    # Filter to find addresses that were selfdestructed but NOT created in the same transaction
+    addresses_to_empty =
+      selfdestruct_addresses
+      |> Enum.reject(fn {tx_hash, address_hash} ->
+        MapSet.member?(created_addresses, {tx_hash, address_hash})
+      end)
+      |> Enum.map(fn {_tx_hash, address_hash} -> address_hash end)
+      |> Enum.uniq()
```
Comment (rimrakhimov) [inline apps/explorer/lib/explorer/chain/import/runner/internal_transactions.ex:939]: Also, we reject the addresses that were self-destructed in the same transaction. But may it happen for the same address to be self-destructed but then created with some other bytecode in the future transactions? What if that future transaction is indexed before the self-destruction transaction, won't we remove the valid bytecode from the future creation transaction?

Upd: it looks like creating a contract with different bytecode is impossible, but if the bytecode remains the same, the `CREATE2` opcode may be used to create the contract at the same address ([link](https://ethereum.stackexchange.com/questions/80388/is-it-possible-to-deploy-a-contract-on-the-same-address-after-self-destruct))
